### ISSUE TITLE
Add a new schema type for representing dimension lists

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#44409]
## 🚀 Feature
Create a specialized schema type that can represent None, int, or tuple of ints. Could be named `Dims`, `DimList`, `ReductionDim`, or something else.

## Motivation

Currently, the only single schema type that can do this is `int[1]?`, which is an `optional<IntArrayRef>` in C++. This type seems to almost work correctly, except that if the function is called from C++ with an integer, the `optional<IntArrayRef>` is filled with a garbage value. The value must be wrapped with `IntArrayRef({<value>})` to get correct behavior:

```
void func(optional<IntArrayRef> opt_dims) {
  if (opt_dims.has_value()) {
    for (auto dim : opt_dims.value()) {
      std::cout << dim << " " << std::endl;
    }
  }
  std::cout << std::endl;
}

func(23); // This prints out a garbage value
func({23}); // This also prints out a garbage value
func(IntArrayRef({23})); // This correctly prints out "23"
```

After `int[n]?` was introduced, I tried to use it to fix #29137, but the issue with `optional<IntArrayRef>` silently breaks any C++ calls that provide an integer as the dim argument. The C++ API of `at::sum` cannot easily be changed to prevent callers from providing an int--for one, that would be a BC break, but also, there does not seem to be any way to detect whether the `optional<IntArrayRef>` argument was initialized in this faulty way.

More details starting here: https://github.com/pytorch/pytorch/pull/43982#issuecomment-689173522
And an older discussion starting here: https://github.com/pytorch/pytorch/pull/30822#issuecomment-571818073

## Pitch

@zdevito wrote a good pitch here: https://github.com/pytorch/pytorch/pull/30822#issuecomment-572783469

If we had a dedicated class to represent an optional list of dimensions, we would avoid having to use `optional<IntArrayRef>`. This dimension list class would have constructors that accept the following:
* An integer
* A list of integers
* A representation of `None` (perhaps c10::nullopt)

## Additional context

I think it's important to consider how this new dimension list class would interact with the `Dimname[n]` schema type, which accepts `None` as a valid value. Currently, if a function has one overload where an argument type is `Dimname[1]`, and another overload where that argument is `int[1]?`, this creates an ambiguous situation since both overloads would accept `None`. Some additional info: https://github.com/pytorch/pytorch/pull/43982#discussion_r481426382

This ambiguous situation compiles without error. Apparently codegen will arbitrarily choose which C++ function gets called (perhaps based on the alphabetical order of the schema overload names). To avoid the ambiguity, we could potentially have something like a `DimnameListNotNone` schema type that does everything that `Dimname[1]` does, except that it does not accept `None`. Perhaps there are other possible solutions though.

@t-vi suggested trying to split `Dimname[1]` into `Dimname?[]` and `Dimname[]`, and I haven't tried that yet. Still, it would not be ideal to have to have two separate overloads for the dimname list case.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @yf225 @glaringlee @bhosmer @smessmer @ljk53 @mruberry @rgommers @heitorschueroff

**Comment by t-vi** [comment]
More precisely, I observed that `None` is a valid Dimname only in some contexts (keeping dimensions unnamed when defining dimensions in factory functions or rename) while not in others (using dimensions e.g. in reductions), so really, we have to types of lists, one of NotNoneDimname and one of DimnameThatCanBeNone (which I'd dubiously call Dimname?[]. rename would get Dimname?[] and sum would get Dimname[1] under this scheme, because [None] isn't a proper set of dimension names for reductions)

**Comment by peterbell10** [comment]
I ran into the `optional<IntArrayRef>` issue again on #50903. The problem is that constructing an `IntArrayRef` from a literal `int` first converts to `int64_t`, followed by storing a pointer to the `int64_t` temporary. When you explicitly construct the `IntArrayRef` as part of the function call, C++ extends the lifetime of the temporary until the entire expression (e.g `func(IntArrayRef({23}))`) is evaluated so there is no dangling pointer. However, when we construct an `optional<IntArrayRef>` from an `int`, the conversion to `int64_t` takes place inside the scope of the `optional` constructor, so the lifetime of the temporary ends _before_ the function call.

IMO the main issue here is that the `ArrayRef(const T&)` constructor is dangerous, especially for fundamental types which are allowed to go through these two-step conversions (`int` -> `int64_t` -> `IntArrayRef`). I would propose either deprecating and removing the overload, or at least making it explicit. Then broadcasting lists would need to be be implemented in codegen for the C++ API by inserting new overloads, e.g.
```cpp
void func(optional<IntArrayRef> opt_dims);

void func(const int64_t single_dim) {
  return func(IntArrayRef(&single_dim, 1));
}
```
Here there is absolutely no possibility to create a danging pointer. It also enforces the distinction between normal list and broadcasting list in the C++ API (currently all lists will accept scalars). That said, this does require a combinatorial number of overloads be generated for signatures with multiple lists. 

Note, there's already a `TODO` to make the constructor `explicit` so presumably someone has thought about this issue before. It just might not have come up until now.
https://github.com/pytorch/pytorch/blob/be51de40473bfbbd77c444aafd785206b9ca2b18/c10/util/ArrayRef.h#L65-L66

**Comment by ezyang** [comment]
cc @smessmer @swolchok

**Comment by kurtamohler** [comment]
Thanks a lot for that explanation, @peterbell10! Nice work finding the cause of the issue

**Comment by ezyang** [comment]
Let's do this! Now that the codegen is rewritten I think it's in a good spot for someone to try doing this.

**Comment by peterbell10** [comment]
By the way, I also have a solution for the `Dimname` issue @t-vi mentioned. See #51350.

Python API generation has a `sort_overloads` step that chooses the order in which python overloads get tried. By prioritising `Dimname` below everything else, the `None` argument will get picked up by the optional overload first. IIUC, this only works when the two overloads have exactly the same type. However, for `Dimname` I think it's reasonable to assume the `int[]?` overload would have the same signature.

**Comment by heitorschueroff** [comment]
Bumping priority as fixing this issue will unblock https://github.com/pytorch/pytorch/issues/61486

**Comment by kurtamohler** [comment]
I can start working on this soon

**Comment by heitorschueroff** [comment]
> I can start working on this soon

Thanks @kurtamohler, let me know how I can help. Just to make sure I'm on the same page here, the fix is not implementing a new type as initially suggested but instead fixing `int[1]?` C++ behavior as suggested by @peterbell10 correct?

**Comment by kurtamohler** [comment]
@heitorschueroff, yep, Peter's suggestion seems like the right way to go

**Comment by heitorschueroff** [comment]
I was trying to better understand the problem so I wrote some minimal code to trigger the same problem, here is a link to cppinsights that shows the template specialization of Wrapper performing a static_cast https://cppinsights.io/s/60f15322.

I'm not an expert on template metaprogramming, but is there a way to prevent template specializations that would require static_casts like these? Or perhaps we can specialize optional for IntArrayRef and disable the constructors for types that can be cast to int64_t?

If there is not way to do the above, then maybe we should actually consider creating a new type to represent a dimension and a list of dimensions (perhaps this new type could encapsulate the logic for named dimensions).

**Comment by ezyang** [comment]
I suppose there are some other tools:
* You can prevent an implicit conversion from occurring by templating. https://stackoverflow.com/a/67005463/23845 
* You can prevent temporaries from being passed into a function by deleting the rvalue reference overloads https://stackoverflow.com/questions/11127807/prevent-pass-by-ref-of-temporary-object

So maybe there is a simpler solution that just involves keeping ArrayRef(const T&) but rejecting ArrayRef(T const &&)

**Comment by ezyang** [comment]
@heitorschueroff and I discussed this and our new plan is to define an OptionalArrayRef<T> that int[]? desugars to, with constructors to prevent the lifetime extension problem

**Comment by kurtamohler** [comment]
My apologies for neglecting this issue, I've been focused on other areas for a while. It looks like @heitorschueroff took over and was working on PR #64084. But that PR is still in draft and hasn't had any activity in a few months. Are you still planning on continuing this @heitorschueroff ?

**Comment by mruberry** [comment]
@heitorschueroff is actually working on networking for Google now, so I think he's probably busy with other things

**Comment by kurtamohler** [comment]
I've drafted a PR that modifies codegen to resolve `int[]?` arguments to the `OptionalIntArrayRef` class that @heitorschueroff drafted in #64084. There are some remaining issues to fix before it should be reviewed. But I have a question.

I'm wondering if its wrong for _all_ `int[]?` arguments to be resolved to `OptionalIntArrayRef`, even if the argument isn't a dimension list. We have some functions with `int[]? size` arguments (and a few others that are not `int[]? dim`), and they too are translated to `OptionalIntArrayRef` in my PR. Would it be better for `int[]? size` to remain `optional<IntArrayRef>` instead? If the answer is yes, then maybe it would be better to create a new schema type for dimension lists after all.

Let me know what you think @ezyang, @mruberry, @peterbell10, and whoever else has some thoughts.

**Comment by peterbell10** [comment]
The dangling pointer issue for scalar arguments should exist for all `int[]?` arguments so I think we do want to change all of them.

**Comment by ezyang** [comment]
concur with peterbell10
